src/memo/memo.mli

```diff
@@ -29,19 +46,78 @@ end
     are forgotten, pending computations are cancelled. *)
 val reset : unit -> unit
 
-module type S = Memo_intf.S with type stack_frame := Stack_frame.t
-module type Input = Memo_intf.Input
-module type Output = Memo_intf.Output
-module type Decoder = Memo_intf.Decoder
+module Function_type : sig
+  type ('a, 'b, 'f) t =
+    | Sync : ('a, 'b, ('a -> 'b)) t
+    | Async : ('a, 'b, ('a -> 'b Fiber.t)) t
+end
 
-module Make(Input : Input)(Decoder : Decoder with type t := Input.t)
-  : S with type input := Input.t
+module type Sexpable = sig
+  type t
+  val to_sexp : t -> Sexp.t
+end
+
+module type Data = sig
+  type t
+  include Table.Key with type t := t
+  include Sexpable with type t := t
+end
+
+(**
+   When we recompute the function and find that its output is the same as what we
+   computed before, we can sometimes skip recomputing the values that depend on it.
+
+   [Allow_cutoff] specifies how to compare the output values for that purpose.
+
+   Note that currently Dune wipes all memoization caches on every run, so
+   cutoff is not effective.
+
+   (* CR-someday aalekseyev:
+   Not sure why the type is so complicated.
+   Couldn't we just pass an equality function as an extra optional argument to [create]?
+   We don't use hash anywhere, I think. *)
+*)
+module Output : sig
+  type 'o t =
+    | Simple of (module Sexpable with type t = 'o)
+    | Allow_cutoff of (module Data with type t = 'o)
+end
+
+module Visibility : sig
+  type 'i t =
+    | Hidden
+    | Public of 'i Dune_lang.Decoder.t
+end
 
-(** Same as [Make] except that registered functions won't be available
-    through [dune compute]. In particular, it is not necessary to
-    provide a decoder for the input. *)
-module Make_hidden(Input : Input)
-  : S with type input := Input.t
+val create
+  :  string
+  -> doc:string
+  -> input:(module Data with type t = 'i)
+  -> visibility:'i Visibility.t
+  -> output:('o Output.t)
+  -> ('i, 'o, 'f) Function_type.t
+  -> 'f option
+  -> ('i, 'o, 'f) t
+
+(** Set the implementation of a memoized function created with
+    [fcreate] *)
```

Comment:
Typo: should be `create`.

Reply:
Thanks. Changed and also added the documentation for `create` that I somehow lost earlier.